libraries/c_sdk/standard/mqtt/src/iot_mqtt_network.c

```diff
@@ -232,11 +232,23 @@ static IotMqttError_t _getIncomingPacket( void * pNetworkConnection,
             EMPTY_ELSE_MARKER;
         }
 
-        dataBytesRead = pMqttConnection->pNetworkInterface->receive( pNetworkConnection,
-                                                                     pIncomingPacket->pRemainingData,
-                                                                     pIncomingPacket->remainingLength );
+        while( totalBytesRead < pIncomingPacket->remainingLength )
+        {
+            bytesRead = pMqttConnection->pNetworkInterface->receive( pNetworkConnection,
+                                                                     pIncomingPacket->pRemainingData + totalBytesRead,
+                                                                     pIncomingPacket->remainingLength - totalBytesRead );
+
+            if( bytesRead == 0 )
```

Comment:
For non blocking sockets, `bytesRead` could be 0, if there is no data currently on socket ( EWOULDBLOCK ).  Previous code handles this within the while loop of `IotNetworkAfr_Receive()`

Reply:
I believe a non-blocking socket returns EWOULDBLOCK while the blocking sockets return 0 on timeout.

Perhaps we need to return socket layers after all, so the application can ask again? 
Suggestion:
```c
/* param[in,out] bytesRequested */
IotNetworkError_t ( * receive )( void * pConnection,
                           uint8_t * pBuffer,
                           size_t* bytesRequested );
```

In the old code, when 0 was returned from SOCKETS_Recv() the function IotNetwokrAfr() did not loop again. It broke out and returned with the possible partial data.

Reply:
My main objection is that the design matches the documentation which states specifically the following:

/**
     * @brief Block and wait for incoming network data.
     *
     * Wait for a message of size `bytesRequested` to arrive on the network and
     * place it in `pBuffer`.
     *
     * @param[in] pConnection The connection to wait on, defined by the network
     * stack.
     * @param[out] pBuffer Where to place the incoming network data. This buffer
     * must be at least `bytesRequested` in size.
     * @param[in] bytesRequested How many bytes to wait for. `pBuffer` must be at
     * least this size.
     *
     * @return The number of bytes successfully received. This should be
     * `bytesRequested` when successful. Any other value may indicate an error.
     */

People (as they should) rely on the fact that the port implements what it promises. Looking at the MQTT code e.g. here seems to prove that this is at least true in MQTT.


       dataBytesRead = pMqttConnection->pNetworkInterface->receive( pNetworkConnection,
                                                                     pIncomingPacket->pRemainingData,
                                                                     pIncomingPacket->remainingLength );

        if( dataBytesRead != pIncomingPacket->remainingLength )
        {
            IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_RESPONSE );
        }
        else
        {
            EMPTY_ELSE_MARKER;
        }

Reply:
This means that you would always have a race condition in MQTT and if you break the packet in half by failing out it looks like bad things will happen after this ...